File: Microbit_joystick/JoyStickTest.js
function bigChange(currValue: number, lastValue: number): number {
    if (currValue - lastValue >= 20 || lastValue - currValue >= 20) {
        return 1
    } else {
        return 0
    }
    
}

bluetooth.onBluetoothConnected(function on_bluetooth_connected() {
    basic.showIcon(IconNames.Yes)
})
bluetooth.onBluetoothDisconnected(function on_bluetooth_disconnected() {
    basic.showIcon(IconNames.No)
})
joystickbit.initJoystickBit()
basic.showIcon(IconNames.No)
bluetooth.startUartService()
basic.pause(1000)
let lastX = 508
let lastY = 508
let currX = 508
let currY = 508
serial.redirectToUSB()
serial.setBaudRate(BaudRate.BaudRate115200)
basic.forever(function on_forever() {
    
    currX = joystickbit.getRockerValue(joystickbit.rockerType.X)
    currY = joystickbit.getRockerValue(joystickbit.rockerType.Y)
    if (joystickbit.getButton(joystickbit.JoystickBitPin.P12)) {
        basic.showString("C")
        bluetooth.uartWriteString("crawl")
    }
    
    if (joystickbit.getButton(joystickbit.JoystickBitPin.P13)) {
        basic.showString("D")
        bluetooth.uartWriteString("walk")
    }
    
    if (joystickbit.getButton(joystickbit.JoystickBitPin.P14)) {
        basic.showString("E")
        bluetooth.uartWriteString("trot")
    }
    
    if (joystickbit.getButton(joystickbit.JoystickBitPin.P15)) {
        basic.showString("F")
        bluetooth.uartWriteString("rest")
    }
    
    if (input.buttonIsPressed(Button.A)) {
        basic.showString("A")
        bluetooth.uartWriteString("hello")
    }
    
    if (input.buttonIsPressed(Button.B)) {
        basic.showString("B")
        bluetooth.uartWriteString("sit")
    }
    
    if (bigChange(currX, lastX) == 1 || bigChange(currY, lastY) == 1) {
        if (currX >= 500 && currX <= 550 && (currY >= 500 && currY <= 550)) {
            basic.pause(800)
            bluetooth.uartWriteString("balance")
            basic.clearScreen()
            basic.pause(200)
        } else {
            if (currX <= 200) {
                bluetooth.uartWriteString("right")
                basic.showLeds(`
                    . . # . .
                    . . . # .
                    # # # # #
                    . . . # .
                    . . # . .
                    `)
            }
            
            if (currX >= 800) {
                bluetooth.uartWriteString("left")
                basic.showLeds(`
                    . . # . .
                    . # . . .
                    # # # # #
                    . # . . .
                    . . # . .
                    `)
            }
            
            if (currY <= 200) {
                serial.writeValue("y", currY)
                bluetooth.uartWriteString("back")
                basic.showLeds(`
                    . . # . .
                    . . # . .
                    # . # . #
                    . # # # .
                    . . # . .
                    `)
            }
            
            if (currY >= 800) {
                bluetooth.uartWriteString("forward")
                basic.showLeds(`
                    . . # . .
                    . # # # .
                    # . # . #
                    . . # . .
                    . . # . .
                    `)
            }
            
            basic.pause(200)
        }
        
        lastX = currX
        lastY = currY
    }
    
})

File: Microbit_joystick/JoyStickTest.py
def bigChange(currValue: number, lastValue: number):
    if currValue - lastValue >= 20 or lastValue - currValue >= 20:
        return 1
    else:
        return 0

def on_bluetooth_connected():
    basic.show_icon(IconNames.YES)
bluetooth.on_bluetooth_connected(on_bluetooth_connected)

def on_bluetooth_disconnected():
    basic.show_icon(IconNames.NO)
bluetooth.on_bluetooth_disconnected(on_bluetooth_disconnected)

joystickbit.init_joystick_bit()
basic.show_icon(IconNames.NO)
bluetooth.start_uart_service()
basic.pause(1000)
lastX = 508
lastY = 508
currX = 508
currY = 508
serial.redirect_to_usb()
serial.set_baud_rate(BaudRate.BAUD_RATE115200)

def on_forever():
    global currX, currY, lastX, lastY
    currX = joystickbit.get_rocker_value(joystickbit.rockerType.X)
    currY = joystickbit.get_rocker_value(joystickbit.rockerType.Y)
    if joystickbit.get_button(joystickbit.JoystickBitPin.P12):
        basic.show_string("C")
        bluetooth.uart_write_string("crawl")
    if joystickbit.get_button(joystickbit.JoystickBitPin.P13):
        basic.show_string("D")
        bluetooth.uart_write_string("walk")
    if joystickbit.get_button(joystickbit.JoystickBitPin.P14):
        basic.show_string("E")
        bluetooth.uart_write_string("trot")
    if joystickbit.get_button(joystickbit.JoystickBitPin.P15):
        basic.show_string("F")
        bluetooth.uart_write_string("rest")
    if input.button_is_pressed(Button.A):
        basic.show_string("A")
        bluetooth.uart_write_string("hello")
    if input.button_is_pressed(Button.B):
        basic.show_string("B")
        bluetooth.uart_write_string("sit")
    if bigChange(currX, lastX) == 1 or bigChange(currY, lastY) == 1:
        if currX >= 500 and currX <= 550 and (currY >= 500 and currY <= 550):
            basic.pause(800)
            bluetooth.uart_write_string("balance")
            basic.clear_screen()
            basic.pause(200)
        else:
            if currX <= 200:
                bluetooth.uart_write_string("right")
                basic.show_leds("""
                    . . # . .
                    . . . # .
                    # # # # #
                    . . . # .
                    . . # . .
                    """)
            if currX >= 800:
                bluetooth.uart_write_string("left")
                basic.show_leds("""
                    . . # . .
                    . # . . .
                    # # # # #
                    . # . . .
                    . . # . .
                    """)
            if currY <= 200:
                serial.write_value("y", currY)
                bluetooth.uart_write_string("back")
                basic.show_leds("""
                    . . # . .
                    . . # . .
                    # . # . #
                    . # # # .
                    . . # . .
                    """)
            if currY >= 800:
                bluetooth.uart_write_string("forward")
                basic.show_leds("""
                    . . # . .
                    . # # # .
                    # . # . #
                    . . # . .
                    . . # . .
                    """)
            basic.pause(200)
        lastX = currX
        lastY = currY
basic.forever(on_forever)
